### ISSUE TITLE
C# Dev Kit Support for Unity Projects

### ISSUE THREAD

**Issue by MahmoudKanbar** [microsoft/vscode-dotnettools#271]
### Describe the feature you'd like

Whenever I try to open a unity project in VSCode using the new C# Dev Kit I get the following warning messages for each project
`[warning] The project file 'PATH TO THE .CSPROJ FILE' is in unsupported format (for example, a traditional .Net Framework project). It need be converted to new SDK style to work in C# Dev Kit. `

![image](https://github.com/microsoft/vscode-dotnettools/assets/39844467/a7dd434b-ebd0-41bc-8b6f-f23fa3fc3e52)

Opening Unity C# projects in the new extension will be very useful for a lot of programmers.
And thanks for the amazing extension.

**Comment by vazab** [comment]
I also start getting an error:
[error] Failed to load project 'path\Assembly-CSharp.csproj'. One or more errors occurred. (This project is not supported in C# Dev Kit.)
This is probably related to .NET Framework 4.8 because it's built into Windows 10. When as the target framework it's 4.7.1 for csproj

**Comment by WagnerWD40** [comment]
Same error here using linux Mint

**Comment by ashtonland** [comment]
Also having this problem using windows 11. Tried installing .NET 4.7.1 SDK and targeting pack to no avail. Seems purely due to the bit older format of unity generated csproj files. I find that a project will succeed at loading when i add the `Sdk="Microsoft.NET.Sdk"` property to Unity's generated `<Project toolsVersion="4.0" ... />` property. I assume this is the "new SDK" style that the error above to which the above error referred. Also I had to add
`<TargetFramework>net6.0</TargetFramework>`  within the property group to eliminate one of the warnings. 

However, while the projects will load, the solution viewer will be useless because, since Unity csproj files are generated in the root directory, each project in the solution viewer just shows the entire project file system like the regular view. This could be fixed oneself by moving your assembly's .csproj file in the folder alongside the assembly definition file, but unity will just regenerate that csproj file at the root directory the next yo you compile your code so that doesn't work either & this regeneration removes the property changes/additions that I stated above which causes the projects to fail to compile once more and back at the beginning...

So evidently I was unable to fix the problem, but I have uncovered why it occurs and what needs to be don't to fix it so hopefully it can help the DevKit team resolve this by making some special treatment for Unity C# Projects :)

**Comment by Facewade** [comment]
So, im super new to unity and got this problem too, the thing is, my intelissense is not working, and i thought this was related, is it?

**Comment by WagnerWD40** [comment]
Yes, that's the cause.

Em seg., 31 de jul. de 2023 20:34, Facewade ***@***.***>
escreveu:

**Comment by damianDAMPortada** [comment]
I suscribre, mine doesn't work either
The project file 'c:\somethingsomething.csproj' is in unsupported format (for example, a traditional .Net Framework project). It need be converted to new SDK style to work in C# Dev Kit.

**Comment by perrylets** [comment]
Not working for me as well. Now it's impossible to use Unity with VS Code a few days after Unity posted a guide on how to debug with VS Code (debugging with unity is different). You can't use one of the most famous use cases of C# with VS Code, great idea.

**Comment by jbevain** [comment]
We just announced the addition of the [Unity extension for Visual Studio Code](https://devblogs.microsoft.com/visualstudio/announcing-the-unity-extension-for-visual-studio-code/) to the C# Dev Kit family of extensions, currently in preview. Make sure to follow the **Getting Started** guide and you should be good to go.

I'm going to close this issue, but feel free to create new issues if you encounter problems with the preview!

**Comment by jbevain** [comment]
@Chizaruu 

> Note that you can't have both packages installed simultaneously for reasons that only Microsoft & Unity know. (See [Chizaruu/com.tsk.ide.vscode#77](https://github.com/Chizaruu/com.tsk.ide.vscode/issues/77) for more details)

Feel free to open a new issue to track this. The behavior and code your highlight in your GitHub issue is simply a way to avoid a conflict between the Visual Studio Code support in the **Visual Studio Editor** Unity Package that we work on in collaboration with Unity, and the **Visual Studio Code Editor** Unity Package that Unity previously maintained and now stopped supporting (but is still installed in many Unity versions).

We'd be happy to look at only disables the legacy and unmaintained Unity Package that Unity still ships if that conflicts with your package.

**Comment by Chizaruu** [comment]
> @Chizaruu 
> 
> > Note that you can't have both packages installed simultaneously for reasons that only Microsoft & Unity know. (See [Chizaruu/com.tsk.ide.vscode#77](https://github.com/Chizaruu/com.tsk.ide.vscode/issues/77) for more details)
> 
> Feel free to open a new issue to track this. The behavior and code your highlight in your GitHub issue is simply a way to avoid a conflict between the Visual Studio Code support in the **Visual Studio Editor** Unity Package that we work on in collaboration with Unity, and the **Visual Studio Code Editor** Unity Package that Unity previously maintained and now stopped supporting (but is still installed in many Unity versions).
> 
> We'd be happy to look at only disabling the legacy and unmaintained Unity Package that Unity still ships if that conflicts with your package.

Sounds like a plan, thanks for letting me know!

**Comment by nguyenminhtri10061997** [comment]
try open Assembly-CSharp.csproj file, add new line value is net sdk you installed before ex:
` <TargetFramework>net7.0</TargetFramework>`
in below
`<TargetFrameworkVersion>v4.7.1</TargetFrameworkVersion>`

**Comment by ajkolenc** [comment]
> > ### Describe the feature you'd like
> > Whenever I try to open a unity project in VSCode using the new C# Dev Kit I get the following warning messages for each project `[warning] The project file 'PATH TO THE .CSPROJ FILE' is in unsupported format (for example, a traditional .Net Framework project). It need be converted to new SDK style to work in C# Dev Kit. `
> > ![image](https://user-images.githubusercontent.com/39844467/254224528-a7dd434b-ebd0-41bc-8b6f-f23fa3fc3e52.png)
> > Opening Unity C# projects in the new extension will be very useful for many programmers. And thanks for the amazing extension.
> > ### Alternatives considered
> > _No response_
> > ### Environment Information
> > _No response_
> 
> You have two options for SDK style csproj.
> 
> 1. Use the new "[officially" supported package](https://devblogs.microsoft.com/visualstudio/announcing-the-unity-extension-for-visual-studio-code/) for VSCode, the standard Visual Studio Package (v2.0.20), on the Unity Package Manager.
> 
> or
> 
> 2. Use my [package ](https://github.com/Chizaruu/com.tsk.ide.vscode)for VSCode support, which generates SDK Style csproj **and** is more community focused.
> 
> Note that you can't have both packages installed simultaneously for reasons that only Microsoft & Unity know. (See [Chizaruu/com.tsk.ide.vscode#77](https://github.com/Chizaruu/com.tsk.ide.vscode/issues/77) for more details)

Upgrading to the latest Visual Studio Package (v2.0.20) fixed this issue for me, I didn't realize that package was involved. Thank you!

**Comment by enjoy-software** [comment]
Follow this document [Unity Development with VS Code](https://code.visualstudio.com/docs/other/unity).
Finally when you reopen the project, unity editor will automatically update .csproj.

**Comment by prakyath-07** [comment]
Delete all CsProj files and soltuion files from the root, and regenerate the files from settings (Edit->Preferences->Regenerate Project Files)

**Comment by daniel-keen** [comment]
Can something please be ever released without any critical issues that block the workflow....
Anyway, I have the same issue as the TS.

**Comment by afiyindev** [comment]
Im having this same issue, ive updated to version 2.0.20 but it still doesnt work

**Comment by Reyder95** [comment]
Having the same issues as well. Downloading the 2.0.20 version of Visual Studio Editor, as well as Visual Studio Code Editor, and regenerating files etc still provides me with 


The project file '...\Assembly-CSharp.csproj' is in unsupported format (for example, a traditional .Net Framework project). It need be converted to new SDK style to work in C# Dev Kit.

Tried pretty much everything here (deleting then regenerating files etc) and nothing seems to work.

**Comment by askor** [comment]
_Removing the Visual Studio Code Editor package_ and updating Visual Studio Editor in Unity Package Manager fixed it for me.

@afiyindev @Reyder95 @daniel-keen

**Comment by Emberstorm-John** [comment]
Complete guide that worked for me: (from answes above)

Make sure you are using Unity 2021 or above
Install VSCode Unity Extension https://marketplace.visualstudio.com/items?itemName=VisualStudioToolsForUnity.vstuc
In Unity go to _Window -> Package Manager_
- Uninstall Visual Studio Code Editor package
- Install and or Update Visual Studio Editor (version 2.0.20+)
In windows delete all .csproj and .sln files from your unity project folder 
In Unity go to _Edit -> Preferences -> External Tools_
- Set VS Code as Code Editor
- Click Regenerate Project Files
In Unity double click any script to open VSCode

**Comment by honzapatCZ** [comment]
> ` <TargetFramework>net7.0</TargetFramework>`

Well, what if you dare to use Unity 2020?
What should one do?

**Comment by XRLabLondon** [comment]
Updated everything, removed Visual Code Editor package in Unity, restarted everything, still no luck...

**Comment by alcidesbava** [comment]
> Having the same issues as well. Downloading the 2.0.20 version of Visual Studio Editor, as well as Visual Studio Code Editor, and regenerating files etc still provides me with
> 
> The project file '...\Assembly-CSharp.csproj' is in unsupported format (for example, a traditional .Net Framework project). It need be converted to new SDK style to work in C# Dev Kit.
> 
> Tried pretty much everything here (deleting then regenerating files etc) and nothing seems to work.


.
.
.

have yousolved the problem?

**Comment by broseidon91** [comment]
> > Having the same issues as well. Downloading the 2.0.20 version of Visual Studio Editor, as well as Visual Studio Code Editor, and regenerating files etc still provides me with
> > The project file '...\Assembly-CSharp.csproj' is in unsupported format (for example, a traditional .Net Framework project). It need be converted to new SDK style to work in C# Dev Kit.
> > Tried pretty much everything here (deleting then regenerating files etc) and nothing seems to work.
> 
> . . .
> 
> have yousolved the problem?

After failing to solve it on this page, i found this: https://discussions.unity.com/t/vscode-cant-load-c-project/273373

Particularly, nothing worked until I added `<TargetFramework>net7.0</TargetFramework>`  in my Assembly-CSharp.csproj file, just under `<TargetFrameworkVersion>v4.7.1</TargetFrameworkVersion>`. After saving, VSCode properly loaded the project.

**Comment by rdvansloten** [comment]
> try open Assembly-CSharp.csproj file, add new line value is net sdk you installed before ex:
> ` <TargetFramework>net7.0</TargetFramework>`
> in below
> `<TargetFrameworkVersion>v4.7.1</TargetFrameworkVersion>`

Oh my god, this actually works, on MacOS at least:

```
<TargetFrameworkVersion>v4.7.1</TargetFrameworkVersion>
<TargetFramework>net7.0</TargetFramework>
```

And the output (after 22:26 is when I changed it and saved)
```
2023-08-13 22:24:43.794 [warning] The project file '/Users/rdvansloten/Unity/Snowboarder01/Assembly-CSharp.csproj' is in unsupported format (for example, a traditional .Net Framework project). It need be converted to new SDK style to work in C# Dev Kit.
2023-08-13 22:24:43.825 [error] Failed to load project '/Users/rdvansloten/Unity/Snowboarder01/Assembly-CSharp.csproj'. One or more errors occurred. (This project is not supported in C# Dev Kit.)
2023-08-13 22:24:43.828 [info] Project system initialization finished. 0 project(s) are loaded, and 1 failed to load.
2023-08-13 22:26:32.892 [warning] (/Users/rdvansloten/Unity/Snowboarder01/Assembly-CSharp.csproj): The reference assemblies for .NETFramework,Version=v4.7.1 were not found. To resolve this, install the Developer Pack (SDK/Targeting Pack) for this framework version or retarget your application. You can download .NET Framework Developer Packs at https://aka.ms/msbuild/developerpacks
2023-08-13 22:26:33.260 [info] Project system initialization finished. 1 project(s) are loaded, and 0 failed to load.
```

**Comment by abelbascu** [comment]
> > > Having the same issues as well. Downloading the 2.0.20 version of Visual Studio Editor, as well as Visual Studio Code Editor, and regenerating files etc still provides me with
> > > The project file '...\Assembly-CSharp.csproj' is in unsupported format (for example, a traditional .Net Framework project). It need be converted to new SDK style to work in C# Dev Kit.
> > > Tried pretty much everything here (deleting then regenerating files etc) and nothing seems to work.
> > 
> > 
> > . . .
> > have yousolved the problem?
> 
> After failing to solve it on this page, i found this: https://discussions.unity.com/t/vscode-cant-load-c-project/273373
> 
> Particularly, nothing worked until I added `<TargetFramework>net7.0</TargetFramework>` in my Assembly-CSharp.csproj file, just under `<TargetFrameworkVersion>v4.7.1</TargetFrameworkVersion>`. After saving, VSCode properly loaded the project.
> 
> For posterity, here are the steps
> 
> 1. Download and install the official version of the unity extension . [Announcing the Unity extension for Visual Studio Code - Visual Studio Blog 137](https://devblogs.microsoft.com/visualstudio/announcing-the-unity-extension-for-visual-studio-code/)
> 2. Go to package manager and unity registry , search “studio” and update both extensions to recommended version.
> 3. Open the VSCode and check for code completion
> 4. If still no code completion , Open the project folder and there you can see all your csproj files for eg : “Assembly-CSharp.csproj”
> 5. Open the csproj file and search for word “TargetFramework”. You will see one like this `<TargetFrameworkVersion>v4.7.1</TargetFrameworkVersion>`
> 6. `<TargetFramework>net7.0</TargetFramework>` copy this and paste under the step 5 line. Do same for all your csproj files if you have more than one.
> 7. Now you can see code completion and syntax highliting etc.
> 
> Since I hadn't coded on my home PC in awhile, I also updated to the latest C# SDK before I did these steps. https://dotnet.microsoft.com/en-us/download

Worked for me too, thanks!!

**Comment by rami-husein** [comment]
FWIW, I think you should add the line ```<TargetFramework>NET_STANDARD_2_1</TargetFramework>``` instead, since that is the API level that unity supports.

**Comment by lalan-desai** [comment]
> try open Assembly-CSharp.csproj file, add new line value is net sdk you installed before ex: ` <TargetFramework>net7.0</TargetFramework>` in below `<TargetFrameworkVersion>v4.7.1</TargetFrameworkVersion>`

Thank you very much it worked for me :)

**Comment by vivekchdry** [comment]
> > ` <TargetFramework>net7.0</TargetFramework>`
> 
> Well, what if you dare to use Unity 2020? What should one do?

Updating to Visual Studio Editor (version 2.0.20)
Worked for me, with using UNity 2020.3.40f1

**Comment by Novack** [comment]
Using Unity 2019.4 here. One silly thing that was making things difficult for me, was not closing and relaunching Unity after making package changes. I asumed this was not required, but finally proved to help.

**Comment by Killerbiene10** [comment]
I've had the same issue, followed all the tips and it went from that error above to this: 

_No installed .NET SDK was found on the computer. If it is installed, make sure it can be found through the PATH environment variable. C# Dev Kit does not work without a supported version of SDK. After installing the SDK, you need restart the VS Code session, or the computer._

<img width="690" alt="Screenshot 2023-09-05 at 10 04 17 PM" src="https://github.com/microsoft/vscode-dotnettools/assets/122064438/6d097514-d913-40b5-8be6-0dd0c64315b6">



I have tried uninstalling, re-installing, setting a path manually, regenerate project files, adding the framework lines up there. 

<img width="579" alt="Screenshot 2023-09-05 at 10 06 11 PM" src="https://github.com/microsoft/vscode-dotnettools/assets/122064438/a1dc37e8-e0f1-4f80-adf5-fd5d1b93e9f0">

I've added the lines:

- Open the csproj file and search for word “TargetFramework”. You will see one like this <TargetFrameworkVersion>v4.7.1</TargetFrameworkVersion>
- <TargetFramework>net7.0</TargetFramework> copy this and paste under the step 5 line. Do same for all your csproj files if you have more than one.

Whenver i reopen Unity the csprod files go back to "<TargetFrameworkVersion>2.0<TargetFrameworkVersion>"I am absolutely at the end of my rope here, completely overwhelmed as to what to do next. I can't work on anything until I have figured this out ...

**Comment by ck9706** [comment]
Maybe can solve this problem:Delete all CsProj files and solution files from the root directory, then regenerate files from Settings (Edit -> Preferences -> Regenerate project files)

**Comment by Killerbiene10** [comment]
Thank you guys - I've solved my issue simply by moving to my PC and using Visual Studio instead. For some reason that works completely fine, and even on my Mac it works much better. VSCode still gets the same errors, but that's a problem for the future, I've wasted enough time on this.

**Comment by LeonSolisPedro** [comment]
I'm also moving to Visual Studio instead, the C# Dev Kit its so buggy for me, and the documentation is so confusing,I wasted so much time trying to make things work.

**Comment by himanshu-nag** [comment]
To resolve the issue of the Unity assembly C# file being in an unsupported format, you'll need to convert it to the new SDK style. Here are the steps to do so: 
 
1. Open your Unity project in Unity Editor. 
2. In the Project window, locate the C# file that is causing the error. 
3. Right-click on the file and select "Open C# Project" or "Open in External IDE" (depending on your Unity version). 
4. This will open the C# project in your preferred Integrated Development Environment (IDE), such as Visual Studio or Visual Studio Code. 
5. In the IDE, right-click on the project file (usually ending with .csproj) and select "Unload Project" or a similar option. 
6. Once the project is unloaded, right-click on the project file again and select "Edit Project File" or "Edit .csproj" (again, the option may vary depending on the IDE). 
7. This will open the project file in a text editor. 
8. Look for the  <Project>  tag and add the following line inside it: 
    <PropertyGroup><LangVersion>latest</LangVersion></PropertyGroup>  
9. Save the changes and close the text editor. 
10. Go back to Unity Editor and right-click on the C# file. 
11. Select "Reload All" or "Reload Scripts" to reload the updated project file. 
12. Unity will now recompile the C# scripts, and the error should be resolved.

**Comment by rootdog404** [comment]
DONT forget to restart Unity!

**Comment by gizzye** [comment]
How do i delete this extension it's malware

**Comment by hdsenevi** [comment]
For me I had to install both `Visual Studio Editor` and `Visual Studio Code Editor` in Unity Package Manager to get this error to go away. Only having `Visual Studio Code Editor` package in Unity package manger game me the error(s) thats discussed above.

![image](https://github.com/microsoft/vscode-dotnettools/assets/886370/145320e8-ad84-4149-9a23-7eda18beca22)

**Comment by rdvansloten** [comment]
For those still struggling with this on Mac (might work for Windows, idk):

- Download .NET 7.0: https://dotnet.microsoft.com/en-us/download and install
- In Visual Studio Code, go to Extensions and install C# Dev Kit, C# and .NET Runtime Install Tool
- In Unity, go to Package Manager and remove "Visual Studio **Code** Editor"
- In Unity, go to Package Manager and make sure "Visual Studio Editor" (not CODE Editor) is installed and version 2.0.21
- Close Visual Studio Code and reopen

**Comment by mcclure** [comment]
> DONT forget to restart Unity after plug-ins updated.

I was struggling horribly with this, and this turned out to be what I was missing. I removed Visual Studio Code Editor, and added Visual Studio Editor\*, and it was continuing to generate the project xml files with the bad `<TargetFrameworkVersion>v4.7.1</TargetFrameworkVersion>` xml no matter how many times I regenerated them. After reading this comment I exited and reopened Unity and did a regenerate and THIS time it generated with a good `<TargetFrameworkVersion>`.

This is very confusing and the error messages are pretty unhelpful! :(

\* Also at this step, after installing Visual Studio Editor and based on a comment on a forum, I went into the Player settings and changed the API Compatibility Level from ".NET Framework" to ".NET Standard 2.1". I do not know if this step did anything at all, but if it did, it had no effect until I restarted Unity.
